ambry-store/src/test/java/com.github.ambry.store/BlobStoreStatsTest.java

```diff
@@ -0,0 +1,335 @@
+import com.github.ambry.utils.Time;
+import com.github.ambry.utils.Utils;
+import com.github.ambry.utils.UtilsTest;
+import java.io.File;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.NavigableMap;
+import java.util.TreeMap;
+import org.junit.After;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+import static org.junit.Assert.*;
+
+
+/**
+ * Tests for {@link PersistentIndex}. Tests both segmented and non segmented log use cases.
+ */
+@RunWith(Parameterized.class)
+public class BlobStoreStatsTest {
+  private static final long TEST_TIME_INTERVAL_IN_MS = CuratedLogIndexState.DELAY_BETWEEN_LAST_MODIFIED_TIMES_MS / 2;
+  private final CuratedLogIndexState state;
+  private final File tempDir;
+  private final BlobStoreStats blobStoreStats;
+
+  /**
+   * Running for both segmented and non-segmented log.
+   * @return an array with both {@code false} and {@code true}.
+   */
+  @Parameterized.Parameters
+  public static List<Object[]> data() {
+    return Arrays.asList(new Object[][]{{false}, {true}});
+  }
+
+  /**
+   * Creates a temporary directory and sets up some test state.
+   * @throws IOException
+   */
+  public BlobStoreStatsTest(boolean isLogSegmented) throws InterruptedException, IOException, StoreException {
+    tempDir = StoreTestUtils.createTempDirectory("indexDir-" + UtilsTest.getRandomString(10));
+    state = new CuratedLogIndexState(isLogSegmented, tempDir);
+    DiskIOScheduler diskIOScheduler = new DiskIOScheduler(null);
+    blobStoreStats = new BlobStoreStats(state.index, state.time, diskIOScheduler);
+  }
+
+  /**
+   * Releases all resources and deletes the temporary directory.
+   * @throws InterruptedException
+   * @throws IOException
+   */
+  @After
+  public void cleanup() throws InterruptedException, IOException, StoreException {
+    state.destroy();
+    assertTrue(tempDir.getAbsolutePath() + " could not be deleted", StoreTestUtils.cleanDirectory(tempDir, true));
+  }
+
+  /**
+   * Basic test to verify reported valid size information per container by BlobStoreStats.
+   * @throws InterruptedException
+   * @throws StoreException
+   * @throws IOException
+   */
+  @Test
+  public void testContainerValidDataSize() throws InterruptedException, StoreException, IOException {
+    verifyContainerValidSize();
+    // advance time
+    state.advanceTime(TEST_TIME_INTERVAL_IN_MS);
+    verifyContainerValidSize();
+  }
+
+  /**
+   * Basic test to verify reported valid size information per log segment by BlobStoreStats.
+   * @throws InterruptedException
+   * @throws StoreException
+   * @throws IOException
+   */
+  @Test
+  public void testLogSegmentValidDataSize() throws InterruptedException, StoreException, IOException {
+    for (long i = 0; i <= state.time.milliseconds() + TEST_TIME_INTERVAL_IN_MS; i += TEST_TIME_INTERVAL_IN_MS) {
```

Comment:
If there is no outlier index segment or index value with a timestamp > state.time then times in the future is the same as asking for the most up-to-date valid data size (now).